engine/access/rpc/backend/backend_events.go

```diff
@@ -100,27 +116,29 @@ func (b *backendEvents) getBlockEventsFromExecutionNode(
 
 // verifyAndConvertToAccessEvents converts execution node api result to access node api result, and verifies that the results contains
 // results from each block that was requested
-func verifyAndConvertToAccessEvents(execEvents []*execproto.GetEventsForBlockIDsResponse_Result, requestedBlockIDs []flow.Identifier) ([]flow.BlockEvents, error) {
-	if len(execEvents) != len(requestedBlockIDs) {
+func verifyAndConvertToAccessEvents(execEvents []*execproto.GetEventsForBlockIDsResponse_Result, requestedBlockHeaders []*flow.Header) ([]flow.BlockEvents, error) {
+	if len(execEvents) != len(requestedBlockHeaders) {
 		return nil, errors.New("number of results does not match number of blocks requested")
 	}
 
-	blockIDSet := map[string]bool{}
-	for _, blockID := range requestedBlockIDs {
-		blockIDSet[blockID.String()] = true
+	reqestedBlockHeaderSet := map[string]*flow.Header{}
+	for _, header := range requestedBlockHeaders {
+		reqestedBlockHeaderSet[header.ID().String()] = header
 	}
 
 	results := make([]flow.BlockEvents, len(execEvents))
 
 	for i, result := range execEvents {
-		if !blockIDSet[hex.EncodeToString(result.GetBlockId())] {
+		header, expected := reqestedBlockHeaderSet[hex.EncodeToString(result.GetBlockId())]
+		if !expected {
 			return nil, fmt.Errorf("unexpected blockID from exe node %x", result.GetBlockId())
 		}
 
 		results[i] = flow.BlockEvents{
-			BlockID:     convert.MessageToIdentifier(result.GetBlockId()),
-			BlockHeight: result.GetBlockHeight(),
-			Events:      convert.MessagesToEvents(result.GetEvents()),
+			BlockID:        convert.MessageToIdentifier(result.GetBlockId()),
+			BlockHeight:    result.GetBlockHeight(),
+			BlockTimestamp: header.Timestamp,
```

Comment:
Is there a reason we're not getting this from the Exe node, and instead filling it here on the Access node?

Reply:
AN already has access to this information available and (to my knowledge) can be trusted when it comes to block information. So there is no need to increase the size of the message going from the EN.

Reply:
I'd almost be tempted to grab all the other block info from the header in that case, and only take the events from the results.

This prevents inconsistencies for cases where the block data on Access node doesn't match that of the Exe node. While this should mostly not happen, taking the block meta info from one specific place is more likely to be consistent.

But no strong preferences here, I think generally the end result will be about the same.

Reply:
I can change `verifyAndConvertToAccessEvents` so It doesn't just that the received block ids match, but also that the that heights match, as extra validation. That way all the data received (from EN) that can be verified (by AN) is verified.

Reply:
done